src/alignment/poa.rs

```diff
@@ -0,0 +1,618 @@
+            let i = node.index() + 1;
+            traceback.last = node;
+            // iterate over the predecessors of this node
+            let prevs: Vec<NodeIndex<usize>> =
+                self.graph.neighbors_directed(node, Incoming).collect();
+            // query base and its index in the DAG (traceback matrix rows)
+            for (j_p, q) in query.iter().enumerate() {
+                let j = j_p + 1;
+                // match and deletion scores for the first reference base
+                let max_cell = if prevs.is_empty() {
+                    TracebackCell {
+                        score: traceback.get(0, j - 1).score + self.scoring.match_fn.score(r, *q),
+                        op: AlignmentOperation::Match(None),
+                    }
+                } else {
+                    let mut max_cell = TracebackCell {
+                        score: MIN_SCORE,
+                        op: AlignmentOperation::Match(None),
+                    };
+                    for prev_node in &prevs {
+                        let i_p: usize = prev_node.index() + 1; // index of previous node
+                        max_cell = max(
+                            max_cell,
+                            max(
+                                TracebackCell {
+                                    score: traceback.get(i_p, j - 1).score
+                                        + self.scoring.match_fn.score(r, *q),
+                                    op: AlignmentOperation::Match(Some((i_p - 1, i - 1))),
+                                },
+                                TracebackCell {
+                                    score: traceback.get(i_p, j).score + self.scoring.gap_open,
+                                    op: AlignmentOperation::Del(Some((i_p - 1, i))),
+                                },
+                            ),
+                        );
+                    }
+                    max_cell
+                };
+
+                let score = max(
+                    max_cell,
+                    TracebackCell {
+                        score: traceback.get(i, j - 1).score + self.scoring.gap_open,
+                        op: AlignmentOperation::Ins(Some(i - 1)),
+                    },
+                );
+                traceback.set(i, j, score);
+            }
+        }
+
+        traceback
+    }
+
+    /// Write the current graph to a specified filepath in dot format for
+    /// visualization, primarily for debugging / diagnostics
+    ///
+    /// # Arguments
+    ///
+    /// * `filename` - The filepath to write the dot file to, as a String
+    ///
+    pub fn write_dot(self, filename: &str) {
+        let mut file = match File::create(&filename) {
+            Err(why) => panic!("couldn't open file {}: {}", filename, why.description()),
+            Ok(file) => file,
+        };
+        let g = self.graph.map(|_, nw| *nw as char, |_, ew| ew);
```

Comment:
Please comment on this mapping.
Also, I wonder if we should rather leave the file writing to the user. Panicing is no good practice, and if we just throw the errors, then there is not much code in this function. I only wonder about the mapping.

Reply:
This method is from the current implementation, I have found the dot representation to be useful for debugging but recently I've been playing with de Bruijn graphs and have found that writing out to GFA files (https://github.com/GFA-spec/GFA-spec) is a nice way to serialize this kind of data -- it even permits including the alignment operations as a cigar string. I see a suggestion to add this was mentioned here: https://github.com/rust-bio/rust-bio/issues/3#issuecomment-260759969

For now I think it's fine to require the user to call something like Dot::new(&poa.graph) explicitly. I agree that we should at least remove the panics and filesystem mutation.

Reply:
Ok, let's remove this method then.